torchdynamo/guards.py

```diff
@@ -309,17 +401,60 @@ def GRAD_MODE(self, guard: Guard):
         """Guard on the current value of torch.is_grad_enabled()"""
         assert guard.name == ""
         assert guard.source is GuardSource.GLOBAL
+        code = None
         if torch.is_grad_enabled():
-            self.code.append("___is_grad_enabled()")
+            code = "___is_grad_enabled()"
         else:
-            self.code.append("not ___is_grad_enabled()")
+            code = "not ___is_grad_enabled()"
+        self._produce_guard_code(guard, [code])
 
     def TENSOR_MATCH(self, guard: Guard):
         if guard.is_nn_module():
             self.ID_MATCH(guard)
         else:
+            value = self.get(guard.name)
             self.tensor_check_names.append(self.arg_ref(guard))
-            self.tensor_check_examples.append(self.get(guard.name))
+            self.tensor_check_examples.append(value)
+
+            # Note: Guard code produced for tensor_match is a little different.
+            # We accumulate tensor names, then do a single install of `___check_tensors`.
+            # See _guards.cpp and TensorGuard for more information.
+            # TODO(voz): Add tensor matching code to export
+            # Note: this is a bit of a special case, and so does not use _produce_guard_code
+            guard.set_export_info(
+                "TENSOR_MATCH", type(value), None, weakref.ref(value), id(value)
+            )
+
+    # A util that appends guarded code, or, in the case of export, adds data onto guards
+    def _produce_guard_code(self, guard, code_list, provided_guarded_object=None):
```

Comment:
self nit, reading this name, it feels a little off, as it doesn't produce guard code so much as add it to existing structures... might rename